infrastructure/src/main/java/org/corfudb/infrastructure/LayoutServer.java

```diff
@@ -141,280 +158,315 @@ private boolean isBootstrapped(CorfuMsg msg, ChannelHandlerContext ctx, IServerR
         final Rank phase1Rank = getPhase1Rank(payloadEpoch);
+        final Rank prepareRank = new Rank(payload.getRank(), getUUID(requestHeader.getClientId()));
 
-        if (payloadEpoch != serverEpoch) {
-            r.sendResponse(ctx, msg, new CorfuPayloadMsg<>(CorfuMsgType.WRONG_EPOCH, serverEpoch));
-            log.trace("handleMessageLayoutPrepare: Incoming message with wrong epoch, got {}, "
-                            + "expected {}, message was: {}",
-                    msg.getPayload().getEpoch(), serverEpoch, msg);
+        if(payloadEpoch != serverEpoch) {
+            r.sendWrongEpochError(requestHeader, ctx, serverEpoch);
             return;
         }
 
-        Layout proposedLayout = getProposedLayout(payloadEpoch);
+        final Layout proposedLayout = getProposedLayout(payloadEpoch);
+        HeaderMsg responseHeader;
+        ResponseMsg response;
+
+        // If the PREPARE_LAYOUT_REQUEST rank is less than or equal to the highest phase 1 rank, reject.
+        if(phase1Rank != null && prepareRank.lessThanEqualTo(phase1Rank)) {
+            log.debug("handlePrepareLayoutRequest[{}]: Rejected phase 1 prepare of rank={}, phase1Rank={}",
+                    requestHeader.getRequestId(), prepareRank, phase1Rank);
 
-        // This is a prepare. If the rank is less than or equal to the phase 1 rank, reject.
-        if (phase1Rank != null && prepareRank.lessThanEqualTo(phase1Rank)) {
-            log.debug("handleMessageLayoutPrepare: Rejected phase 1 prepare of rank={}, "
-                    + "phase1Rank={}", prepareRank, phase1Rank);
-            r.sendResponse(ctx, msg, CorfuMsgType.LAYOUT_PREPARE_REJECT.payloadMsg(new
-                    LayoutPrepareResponse(phase1Rank.getRank(), proposedLayout)));
+            responseHeader = getHeaderMsg(requestHeader, false, false);
+            response = getResponseMsg(responseHeader, getPrepareLayoutResponseMsg(
+                    PrepareLayoutResponseMsg.Type.REJECT, phase1Rank.getRank(), proposedLayout));
         } else {
             // Return the layout with the highest rank proposed before.
-            Rank highestProposedRank = proposedLayout == null ? new Rank(-1L, msg.getClientID())
-                    : getPhase2Rank(payloadEpoch);
+            Rank highestProposedRank = proposedLayout == null ?
+                    new Rank(-1L, getUUID(requestHeader.getClientId())) : getPhase2Rank(payloadEpoch);
+
             setPhase1Rank(prepareRank, payloadEpoch);
-            log.debug("handleMessageLayoutPrepare: New phase 1 rank={}", prepareRank);
-            r.sendResponse(ctx, msg, CorfuMsgType.LAYOUT_PREPARE_ACK.payloadMsg(new
-                    LayoutPrepareResponse(highestProposedRank.getRank(), proposedLayout)));
+            log.debug("handlePrepareLayoutRequest[{}]: New phase 1 rank={}", requestHeader.getRequestId(), prepareRank);
+
+            responseHeader = getHeaderMsg(requestHeader, false, true);
```

Comment:
Why is this `responseHeader` different than the one in line 277

The response header should always copy its metadata from the requestHeader.

Reply:
All of the fields are the same, except possibly the `ignoreEpoch `or `ignoreClusterId ` fields. These differ because the message in 277 corresponds to a NACK/REJECT while here it corresponds to an ACK. Note that in the original `CorfuMsgType ` implementation, an ACK will ignore the epoch check while a NACK does not.